### ISSUE TITLE
await all async calls assigned to var await_var_assignment

### ISSUE THREAD

**Issue by bsutton** [dart-lang/linter#4378]
## Description

In an async method, any calls to an async method that are assigned to a 'var' must be awaited.

## Details
```dart

Future<bool?> myasyncMethod() => Future.value(true);

void main() async {
    var isTrue = myAsyncMethod();

    Future<bool> isAlsoTrue = myAsyncMethod();
}
```

The above code would generate a lint error on the first call to myAsyncMethod but not the second call.

`Async assignments to var must be awaited. Use an explicit declaration Future<bool> if you don't want to await the assignment`

## Kind

This avoids errors where a user has forgotten to await a call.
Pairing this lint with a change to the formatter will also remove the burden of having manually await every call as the formatter could automatically add the await.

It also stops the formatted removing code.

e.g.
```dart

if (myAsyncMethod() == null) 
{
       // do something
}
```
If the myAsyncMethod was originally a sync method and you modify it to an async method then the formatted
will remove the if statement (as the return type is no longer nullable).

I got caught by this and only picked it up during the review (more by luck than good planning).

## Bad Examples

```dart

Future<bool?> myasyncMethod() => Future.value(true);

void main() async {
   /// bad: fire a lint error
    var isTrue = myAsyncMethod();
}
```

## Good Examples 

```dart

Future<bool?> myasyncMethod() => Future.value(true);

void main() async {
   // good 
    var isTrue = myAsyncMethod();
}
```

```dart

Future<bool?> myasyncMethod() => Future.value(true);

void main() async {
// good no lint will fire.
    Future<bool> isAlsoTrue = myAsyncMethod();
}
```
## Discussion

This lint is an example of ensuring that the developer's intent is clear.

I do wonder if we should change the 'unawaited' method to return a future.

You could then write:

```dart

Future<bool?> myasyncMethod() => Future.value(true);

void main() async {
// good no lint will fire.
   var  isAlsoTrue = unawaited(myAsyncMethod());
}
```
This wouldn't trigger the lint and again make the coder's intent explicit.

### Discussion checklist

- [ ] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [ ] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [ ] If there's any prior art (e.g., in other linters), please add references here.
- [ ] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [ ] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by srawlins** [comment]
Can you give some examples of what bad code this rule would prevent? I ask because in your examples above, the inferred type of the assignment target would be `Future<Something>`, instead of `Something`. It seems like the error would become quite clear as soon as you tried to use the assignment target as a `Foo` instead of as a `Future`.

If there were no false positives to this idea, I think we would just amend `unawaited_futures` to report assignments. The test cases for `unawaited_futures` specifically calls out that assignments don't need to be awaited. Perhaps you want to call the function, do some other work, and then await the result, or just pass it along somewhere else, before awaiting it.

**Comment by bsutton** [comment]
The core issue is about the speed of development.

If I change an existing sync method to async it creates a cascade that I have to go and manually fix.
The other scenario is where I increase the lints and find lots of call sites where the dev has made async calls in sync methods. Fixing these call sites is tedious.
I've often had to go through code and change hundreds of call sites.

This lint would allow the formatter/fix to automatically add awaits rather than me manually having to do it.

Possible error:
There is also a potential error when a Future method overlaps that of the returned class

```dart
void main(List<String> args) async {
  final a = A();
  final something = a.value('ho');

  final sb = StringBuffer();
  // writes the incorrect string to the buffer.
  sb.write(something.toString());
}

class A {
  Future<String> value(String i) => Future.value('hi');
}
```
I've actually done this in real life.


> I think we would just amend unawaited_futures to report assignments
I think this would require the 'unawaited' method to return a future so you can silence the linter when you do actually want the future or as notes silence the lint when the type is explicitly declared as Future.

I would suggest the most common use case is that an async call is awaited so the lint should lean in that direction, i.e explicitly typing the Future if you don't want to await the code and don't want to use unawaited.

The intent here is to reduce the amounts of times that I need to type await and eliminate the possible error noted above.
This may need to be a separate lint as I suspect that an element of the community will be resistant to the idea of having to explicitly type the future or use unawaited.  A separate lint would allow those that want it, to adopt it.

**Comment by srawlins** [comment]
> This lint would allow the formatter/fix to automatically add awaits rather than me manually having to do it.

CC @bwilkerson I see an alternative that might be quite functional. We could offer an assist (no lint needs to be reported) on a function body that inserts `await`s for every Function-returning method invocation.

The `toString` example you give is a good one. If you write `var thisIsntMeantToBeAFuture = ...` and then only call `toString`, `hashCode`, `==`, or `runtimeType` on it, you won't see your mistake. I thought there was a lint rule request, but I can't find it now, to report any `e.toString()` or `"$e"` where `e` is a Future. I think this would be a high value rule.

**Comment by bsutton** [comment]
> We could offer an assist (no lint needs to be reported) on a function body that inserts awaits for every Function

I think it would be far more effective if it was coupled with a lint.

You are notified of a possible problem and the fix is on the line with the problem.

**Comment by bwilkerson** [comment]
> We could offer an assist ...

The downside of an assist is that it can only be run manually. We've talked about ways to make it possible for `dart fix` to run assists, but the needed support doesn't yet exist.

> If I change an existing sync method to async it creates a cascade that I have to go and manually fix.

I have to wonder whether it would be useful to have a refactoring operation for this. We have an assist that makes the local changes, but we could potentially write a refactoring that would
- update overridden and overriding methods where necessary, and
- find all of the call sites and recursively convert the containing method to also be async.

I don't know
- how often users perform this operation,
- how often the behavior of the refactoring would match what users intend to do,
- whether that would negate the need for the requested lint, or
- whether we have time in our schedule to implement such a refactoring.

**Comment by lrhn** [comment]
Is this just for calls, or is it a problem every time a local variable is inferred to a `Future` type?

(Which means that if you *do* want a future-typed local variable, you need to declare it as such, and risk the wrath of the "types on the right" lint)


> I think this would require the 'unawaited' method to return a future

That would make it not work. The `unawaited` function isn't special, the lint doesn't know about it.
It's just a function which takes a future as argument and does *not* return a future. That makes the unawaited future "go away", because passing a future as argument does not consider the future unused. Any function declared as `void foo(Future<Object?> _){}` could be used instead,.

**Comment by bsutton** [comment]
> Is this just for calls, or is it a problem every time a local variable is inferred to a Future type?

I think the issue is around calls.

I don't recall encountering a problem where I assigned a future, this is probably because the problem becomes evident at the site where the original future is created.

>  The unawaited function isn't special, the lint doesn't know about it.

Is there a reason why we couldn't make the lint aware of the method?

Alternatively:

I've often wondered if the 'unawaited' method should be a keyword, but I assume it's to late to consider that given the name clash. - 'unawait' maybe?

The unawait keyword would be a noop but clearly declares the developer's intent.
Having said that, I am a little uncomfortable with the idea of introducing a keyword that implements no actual functionality.  It is however more elegant than having to silence the linter with a comment.

**Comment by lrhn** [comment]
> Is there a reason why we couldn't make the lint aware of the method?

No, but it has not been necessary for it to know. Any function which takes a `Future` as argument is a valid way to dispose of a future, since it will presumably do something with that future. The `unawaited` function is just there to provide a way to deliberately and explicitly do *nothing* with a future. There is no safety in that, only documentation.

If `unawaited` returned the future again, but was magically recognized by the lint as meaning that *that* particular future doesn't need to be awaited ... then we suddenly need to track the value using flow analysis, to recognize whether an expression of type `Future` has a value which definitely comes out of a call to `unawaited`.
That's much more complicated than just having `unawaited` return `void`, and complain about *every* expression with static type `Future<Something>` in a void context.

(Another good alternative to ignore futures is to call `future.ignore();`, which does nothing with the future *and* prevents any errors from becoming unhandled.)

> I've often wondered if the 'unawaited' method should be a keyword, but I assume it's to late to consider that given the name clash. - 'unawait' maybe?

The language is blissfully unaware that there is a lint telling you to not ignore futures. It doesn't care if you do, the semantics are clear either way. (Well, especially since `unwaited` does nothing, so it's the same semantics.)
I consider it very unlikely that we'd add a *language feature* to account for, from the language's perspective, well ... nothing.

The only reason to do that would be that "unawaited futures" becomes a language feature too, that you *must not* have a `Future`-typed expression in a `void` context (where a `void`-typed expression is allowed). Every future must be awaited. (That would probably carry over to non-`async` functions too, otherwise it's just half a feature.)
And even then, the existing `unwaited` function still provides a workaround to that, no language feature required (other than not rejecting `unawaited` itself for ignoring a future-typed variable), and as you say, it would be an operator with no effect, other than to silently discard a future.

(The necessary steps to ensure that no future is *ever* lost are both interesting to think about, and prohibitively complex. It would effectively force static-single-assignment of every future-typed variable, because assigning a future to a variable, then overwriting the value without using the future, would also be a way to lose a future. We'd have to introduce "definitely read" variables, just like "definitely assigned", where it's an error to end a function with a potentially-future-typed potentially-assigned variable which is not also definitely read, or write to that variable. If that was the goal, I wouldna start from here.)
